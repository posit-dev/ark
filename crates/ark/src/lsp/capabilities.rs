//
// capabilities.rs
//
// Copyright (C) 2025 Posit Software, PBC. All rights reserved.
//
//

use tower_lsp::lsp_types;
use tower_lsp::lsp_types::CodeActionKind;
use tower_lsp::lsp_types::CodeActionOptions;
use tower_lsp::lsp_types::CodeActionProviderCapability;
use tower_lsp::lsp_types::WorkDoneProgressOptions;

/// Capabilities negotiated with [lsp_types::ClientCapabilities]
#[derive(Debug)]
pub(crate) struct Capabilities {
    dynamic_registration_for_did_change_configuration: bool,
    code_action_literal_support: bool,
    workspace_edit_document_changes: bool,
}

impl Capabilities {
    pub(crate) fn new(client_capabilities: lsp_types::ClientCapabilities) -> Self {
        let dynamic_registration_for_did_change_configuration = client_capabilities
            .workspace
            .as_ref()
            .and_then(|workspace| workspace.did_change_configuration)
            .and_then(|did_change_configuration| did_change_configuration.dynamic_registration)
            .unwrap_or(false);

        // In theory the client also tells us which code action kinds it supports inside
        // `code_action_literal_support`, but clients are guaranteed to ignore any they
        // don't support, so we just return `true` if the field exists (same as
        // rust-analyzer).
        let code_action_literal_support = client_capabilities
            .text_document
            .as_ref()
            .and_then(|text_document| text_document.code_action.as_ref())
            .and_then(|code_action| code_action.code_action_literal_support.as_ref())
            .map_or(false, |_| true);

        let workspace_edit_document_changes = client_capabilities
            .workspace
            .as_ref()
            .and_then(|workspace| workspace.workspace_edit.as_ref())
            .and_then(|workspace_edit| workspace_edit.document_changes)
            .map_or(false, |document_changes| document_changes);

        Self {
            dynamic_registration_for_did_change_configuration,
            code_action_literal_support,
            workspace_edit_document_changes,
        }
    }

    pub(crate) fn dynamic_registration_for_did_change_configuration(&self) -> bool {
        self.dynamic_registration_for_did_change_configuration
    }

    pub(crate) fn code_action_literal_support(&self) -> bool {
        self.code_action_literal_support
    }

    // Currently only used for testing
    #[cfg(test)]
    pub(crate) fn with_code_action_literal_support(
        mut self,
        code_action_literal_support: bool,
    ) -> Self {
        self.code_action_literal_support = code_action_literal_support;
        return self;
    }

    pub(crate) fn workspace_edit_document_changes(&self) -> bool {
        self.workspace_edit_document_changes
    }

    // Currently only used for testing
    #[cfg(test)]
    pub(crate) fn with_workspace_edit_document_changes(
        mut self,
        workspace_edit_document_changes: bool,
    ) -> Self {
        self.workspace_edit_document_changes = workspace_edit_document_changes;
        return self;
    }

    pub(crate) fn code_action_provider_capability(&self) -> Option<CodeActionProviderCapability> {
        if !self.code_action_literal_support() {
            return None;
        }

        // Currently we only support documentation generating code actions, which don't
        // map to an existing kind. rust-analyzer maps them to `EMPTY`, so we follow suit.
        // Currently no code actions require delayed resolution.
        Some(CodeActionProviderCapability::Options(CodeActionOptions {
            code_action_kinds: Some(vec![CodeActionKind::EMPTY]),
            work_done_progress_options: WorkDoneProgressOptions::default(),
            resolve_provider: Some(false),
        }))
    }
}

// This is unfortunately required right now, because `LspState` is initialized before we
// get the `Initialize` LSP request. We immediately overwrite the `LspState`
// `capabilities` field after receiving the `Initialize` request.
impl Default for Capabilities {
    fn default() -> Self {
        Self {
            dynamic_registration_for_did_change_configuration: false,
            code_action_literal_support: false,
            workspace_edit_document_changes: false,
        }
    }
}
