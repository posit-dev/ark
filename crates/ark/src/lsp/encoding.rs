//
// encoding.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use ropey::Rope;
use tower_lsp::lsp_types::Position;
use tree_sitter::Point;

/// `PositionEncodingKind` describes the encoding used for the `Position` `character`
/// column offset field. The `Position` `line` field is encoding agnostic, but the
/// `character` field specifies the number of characters offset from the beginning of
/// the line, and the "character" size is dependent on the encoding. The LSP specification
/// states:
///
/// - UTF8: Character offsets count UTF-8 code units (e.g. bytes).
/// - UTF16: Character offsets count UTF-16 code units (default).
/// - UTF32: Character offsets count UTF-32 code units (these are the same as Unicode
///   codepoints, so this `PositionEncodingKind` may also be used for an encoding-agnostic
///   representation of character offsets.)
///
/// The `vscode-languageclient` library that Positron uses on the frontend to create the
/// `Client` side of the LSP currently ONLY supports `UTF16`, and will error on anything
/// else. Their reasoning is that it is easier for the server (ark) to do the re-encoding,
/// since we are tracking the full document state. Track support for UTF-8 here:
/// https://github.com/microsoft/vscode-languageserver-node/issues/1224
///
/// The other interesting part of this is that `TextDocumentContentChangeEvent`s that
/// come through the `did_change()` event and the `TextDocumentItem` that comes through
/// the `did_open()` event encode the `text` of the change/document in UTF-8, even though
/// the `Range` (in the case of `did_change()`) that tells you where to apply the change
/// uses UTF-16, so that's cool. UTF-8 `text` is forced to come through due to how the
/// LSP specification uses jsonrpc, where the content fields must be 'utf-8' encoded:
/// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#contentPart
/// This at least means we have a guarantee that the document itself and any updates to
/// it will be encoded in UTF-8, even if the Positions are UTF-16.
///
/// So we need a way to convert the UTF-16 `Position`s to UTF-8 `tree_sitter::Point`s and
/// back. This requires the document itself, and is what the helpers in this file implement.
pub fn get_position_encoding_kind() -> tower_lsp::lsp_types::PositionEncodingKind {
    tower_lsp::lsp_types::PositionEncodingKind::UTF16
}

pub fn convert_tree_sitter_range_to_lsp_range(
    x: &Rope,
    range: tree_sitter::Range,
) -> tower_lsp::lsp_types::Range {
    let start = convert_point_to_position(x, range.start_point);
    let end = convert_point_to_position(x, range.end_point);
    tower_lsp::lsp_types::Range::new(start, end)
}

pub fn convert_position_to_point(x: &Rope, position: Position) -> Point {
    let line = position.line as usize;
    let character = position.character as usize;

    let character = with_line(x, line, character, convert_character_from_utf16_to_utf8);

    Point::new(line, character)
}

pub fn convert_point_to_position(x: &Rope, point: Point) -> Position {
    let line = point.row;
    let character = point.column;

    let character = with_line(x, line, character, convert_character_from_utf8_to_utf16);

    let line = line as u32;
    let character = character as u32;

    Position::new(line, character)
}

fn with_line<F>(x: &Rope, line: usize, character: usize, f: F) -> usize
where
    F: FnOnce(&str, usize) -> usize,
{
    let Some(x) = x.get_line(line) else {
        let n = x.len_lines();
        let x = x.to_string();
        let line = line + 1;
        // Forcing a full capture so we can learn the situations in which this occurs
        let trace = std::backtrace::Backtrace::force_capture();
        log::error!(
            "Requesting line {line} but only {n} lines exist.\n\nDocument:\n{x}\n\nBacktrace:\n{trace}"
        );
        return 0;
    };

    // If the line is fully contained in a single chunk (likely is), use free conversion to `&str`
    if let Some(x) = x.as_str() {
        return f(x, character);
    }

    // Otherwise, use ever so slightly more expensive String materialization of the
    // line spread across chunks
    let x = x.to_string();
    let x = x.as_str();

    f(x, character)
}

/// Converts a character offset into a particular line from UTF-16 to UTF-8
fn convert_character_from_utf16_to_utf8(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // Initial check, since loop would skip this case
    if character == 0 {
        return character;
    }

    let mut n = 0;

    // For each `u32` sized `char`, figure out the equivalent size in UTF-16
    // world of that `char`. Once we hit the requested number of `character`s,
    // that means we have indexed into `x` to the correct position, at which
    // point we can take the current bytes based `pos` that marks the start of
    // this `char`, and add on its UTF-8 based size to return an adjusted column
    // offset. We use `==` because I'm fairly certain they should always align
    // exactly, and it would be good to log if that isn't the case.
    for (pos, char) in x.char_indices() {
        n += char.len_utf16();

        if n == character {
            return pos + char.len_utf8();
        }
    }

    log::error!("Failed to locate UTF-16 offset of {character}. Line: '{x}'.");
    return 0;
}

/// Converts a character offset into a particular line from UTF-8 to UTF-16
fn convert_character_from_utf8_to_utf16(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // The UTF-8 -> UTF-16 case is slightly simpler. We just slice into `x`
    // using our existing UTF-8 offset, reencode the slice as a UTF-16 based
    // iterator, and count up the pieces.
    match x.get(..character) {
        Some(x) => x.encode_utf16().count(),
        None => {
            let n = x.len();
            log::error!(
                "Tried to take UTF-8 character {character}, but only {n} characters exist. Line: '{x}'."
            );
            0
        },
    }
}
