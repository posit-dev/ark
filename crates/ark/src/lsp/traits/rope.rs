// 
// rope.rs
// 
// Copyright (C) 2022 by RStudio, PBC
// 
// 

use ropey::Rope;
use tower_lsp::lsp_types::Position;
use tree_sitter::Point;

pub(crate) trait RopeExt {
    fn point_to_byte(&self, point: Point) -> usize;
    fn position_to_byte(&self, position: Position) -> usize;
}

impl RopeExt for Rope {

    // TODO: We likely need to translate column positions into byte positions,
    // to properly handle multibyte unicode characters.
    fn point_to_byte(&self, point: Point) -> usize {
        self.line_to_byte(point.row) + point.column
    }

    fn position_to_byte(&self, position: Position) -> usize {
        self.line_to_byte(position.line as usize) + position.character as usize
    }

}
