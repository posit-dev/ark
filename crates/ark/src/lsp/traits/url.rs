//
// url.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use anyhow::*;
use stdext::unwrap;
use std::path::PathBuf;
use std::result::Result::Ok;
use tower_lsp::lsp_types::Url;

pub trait UrlExt {
    fn file_path(&self) -> anyhow::Result<PathBuf>;
}

impl UrlExt for Url {

    fn file_path(&self) -> anyhow::Result<PathBuf> {

        let pathbuf = unwrap!(self.to_file_path(), Err(_) => {
            return Err(anyhow!("error converting URI {} to PathBuf", self));
        });

        Ok(pathbuf)

    }
}
