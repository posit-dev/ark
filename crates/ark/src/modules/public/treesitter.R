
#
# treesitter.R
#
# Copyright (C) 2023 Posit Software, PBC. All rights reserved.
#
#

.ps.treesitter.node.text <- function(node, contents) {
    .ps.Call("ps_treesitter_node_text", node, contents)
}

.ps.treesitter.node.kind <- function(node) {
    .ps.Call("ps_treesitter_node_kind", node)
}
