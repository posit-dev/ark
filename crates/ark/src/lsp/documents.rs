//
// document.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use std::sync::Arc;

use anyhow::*;
use dashmap::DashMap;
use lazy_static::lazy_static;
use ropey::Rope;
use tower_lsp::lsp_types::DidChangeTextDocumentParams;
use tower_lsp::lsp_types::TextDocumentContentChangeEvent;
use tower_lsp::lsp_types::Url;
use tree_sitter::InputEdit;
use tree_sitter::Parser;
use tree_sitter::Point;
use tree_sitter::Tree;

use crate::lsp::traits::position::PositionExt;
use crate::lsp::traits::rope::RopeExt;

lazy_static! {

    // The document index. Stored as a global since various components need to
    // access the document index, and we want to do so without needing to share
    // too many pieces everywhere.
    //
    // Note that DashMap uses synchronization primitives internally, so we
    // don't guard access to the map via a mutex.
    pub static ref DOCUMENT_INDEX: Arc<DashMap<Url, Document>> = Default::default();

}

fn compute_point(
    point: Point,
    text: &str,
) -> Point {
    // figure out where the newlines in this edit are
    let newline_indices: Vec<_> = text.match_indices('\n').collect();
    let num_newlines = newline_indices.len();
    let num_bytes = text.as_bytes().len();

    if newline_indices.len() == 0 {
        return Point::new(point.row, point.column + num_bytes);
    } else {
        let last_newline_index = newline_indices.last().unwrap();
        return Point::new(
            point.row + num_newlines,
            num_bytes - last_newline_index.0 - 1,
        );
    }
}

pub struct Document {
    // The document's textual contents.
    pub contents: Rope,

    // A set of pending changes for this document.
    pub pending: Vec<DidChangeTextDocumentParams>,

    // The version of the document we last synchronized with.
    // None if the document hasn't been synchronized yet.
    pub version: Option<i32>,

    // The parser used to generate the AST.
    pub parser: Parser,

    // The document's AST.
    pub ast: Tree,
}

impl std::fmt::Debug for Document {
    fn fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
    ) -> std::fmt::Result {
        f.debug_struct("Document")
            .field("contents", &self.contents)
            .field("ast", &self.ast)
            .finish()
    }
}

impl Document {
    pub fn new(contents: &str) -> Self {
        // create initial document from rope
        let document = Rope::from(contents);

        // create a parser for this document
        let mut parser = Parser::new();
        parser
            .set_language(tree_sitter_r::language())
            .expect("failed to create parser");
        let ast = parser.parse(contents, None).unwrap();

        let pending = Vec::new();
        let version = None;

        // return generated document
        Self {
            contents: document,
            pending,
            version,
            parser,
            ast,
        }
    }

    pub fn on_did_change(
        &mut self,
        params: &DidChangeTextDocumentParams,
    ) -> Result<i32> {
        // Add pending changes.
        self.pending.push(params.clone());

        // Check the version of this update.
        //
        // If we receive version {n + 2} before {n + 1}, then we'll
        // bail here, and handle the {n + 2} change after we received
        // version {n + 1}.
        //
        // TODO: What if an intermediate document change is somehow dropped or lost?
        // Do we need a way to recover (e.g. reset the document state)?
        if let Some(old_version) = self.version {
            let new_version = params.text_document.version;
            if new_version > old_version + 1 {
                log::info!("on_did_change(): received out-of-order document changes; deferring");
                return Ok(old_version);
            }
        }

        // Get pending updates, sort by version, and then apply them one-by-one.
        self.pending.sort_by(|lhs, rhs| {
            let lhs = lhs.text_document.version;
            let rhs = rhs.text_document.version;
            lhs.cmp(&rhs)
        });

        // Get the maximum version.
        let version = self.pending.last().unwrap().text_document.version;

        // Take the changes, and apply them one-by-one.
        let changes = std::mem::take(&mut self.pending);
        for change in changes {
            for event in change.content_changes {
                if let Err(error) = self.update(&event) {
                    log::error!("error updating document: {}", error);
                }
            }
        }

        // Updates successfully applied; update cached document version.
        self.version = Some(version);

        Ok(version)
    }

    fn update(
        &mut self,
        change: &TextDocumentContentChangeEvent,
    ) -> Result<()> {
        // Extract edit range. Nothing to do if there wasn't an edit.
        let range = match change.range {
            Some(r) => r,
            None => return Ok(()),
        };

        // Update the AST. We do this before updating the underlying document
        // contents, because edit computations need to be done using the current
        // state of the document (prior to the edit being applied) so that byte
        // offsets can be computed correctly.
        let ast = &mut self.ast;

        let start_byte = self.contents.position_to_byte(range.start);
        let start_position = range.start.as_point();

        let old_end_byte = self.contents.position_to_byte(range.end);
        let new_end_byte = start_byte + change.text.as_bytes().len();

        let old_end_position = range.end.as_point();
        let new_end_position = compute_point(start_position, &change.text);

        let edit = InputEdit {
            start_byte,
            old_end_byte,
            new_end_byte,
            start_position,
            old_end_position,
            new_end_position,
        };

        ast.edit(&edit);

        // Now, apply edits to the underlying document.
        let lhs = self.contents.position_to_byte(range.start);
        let rhs = self.contents.position_to_byte(range.end);

        // Now, convert from byte offsets to character offsets.
        let lhs = self.contents.byte_to_char(lhs);
        let rhs = self.contents.byte_to_char(rhs);

        // Remove the old slice of text, and insert the new slice of text.
        self.contents.remove(lhs..rhs);
        self.contents.insert(lhs, change.text.as_str());

        // We've edited the AST, and updated the document. We can now re-parse.
        let contents = self.contents.to_string();
        let ast = self.parser.parse(&contents, Some(&self.ast));
        self.ast = ast.unwrap();

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_point_computation() {
        // empty strings shouldn't do anything
        let point = compute_point(Point::new(0, 0), "");
        assert_eq!(point, Point::new(0, 0));

        let point = compute_point(Point::new(42, 42), "");
        assert_eq!(point, Point::new(42, 42));

        // text insertion without newlines should just extend the column position
        let point = compute_point(Point::new(0, 0), "abcdef");
        assert_eq!(point, Point::new(0, 6));

        // text insertion with newlines should change the row
        let point = compute_point(Point::new(0, 0), "abc\ndef\nghi");
        assert_eq!(point, Point::new(2, 3));

        let point = compute_point(Point::new(0, 0), "abcdefghi\n");
        assert_eq!(point, Point::new(1, 0));
    }
}
