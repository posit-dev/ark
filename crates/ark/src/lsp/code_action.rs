use std::collections::HashMap;

use tower_lsp::lsp_types::CodeAction;
use tower_lsp::lsp_types::CodeActionKind;
use tower_lsp::lsp_types::CodeActionOrCommand;
use tower_lsp::lsp_types::CodeActionResponse;
use tower_lsp::lsp_types::DocumentChanges;
use tower_lsp::lsp_types::OneOf;
use tower_lsp::lsp_types::OptionalVersionedTextDocumentIdentifier;
use tower_lsp::lsp_types::TextDocumentEdit;
use tower_lsp::lsp_types::TextEdit;
use tower_lsp::lsp_types::WorkspaceEdit;
use tree_sitter::Range;
use url::Url;

use crate::lsp::capabilities::Capabilities;
use crate::lsp::code_action::roxygen::roxygen_documentation;
use crate::lsp::documents::Document;

mod roxygen;

/// A small wrapper around [CodeActionResponse] that make a few things more ergonomic
pub(crate) struct CodeActions {
    response: CodeActionResponse,
}

pub(crate) fn code_actions(
    uri: &Url,
    document: &Document,
    range: Range,
    capabilities: &Capabilities,
) -> CodeActionResponse {
    let mut actions = CodeActions::new();

    roxygen_documentation(&mut actions, uri, document, range, capabilities);

    actions.into_response()
}

pub(crate) fn code_action(title: String, kind: CodeActionKind, edit: WorkspaceEdit) -> CodeAction {
    CodeAction {
        title,
        kind: Some(kind),
        edit: Some(edit),
        diagnostics: None,
        command: None,
        is_preferred: None,
        disabled: None,
        data: None,
    }
}

/// Creates a common kind of `WorkspaceEdit` composed of one or more `TextEdit`s to
/// apply to a single document
pub(crate) fn code_action_workspace_text_edit(
    uri: Url,
    version: Option<i32>,
    edits: Vec<TextEdit>,
    capabilities: &Capabilities,
) -> WorkspaceEdit {
    if capabilities.workspace_edit_document_changes() {
        // Prefer the versioned `DocumentChanges` feature
        let edit = TextDocumentEdit {
            text_document: OptionalVersionedTextDocumentIdentifier { uri, version },
            edits: edits.into_iter().map(|edit| OneOf::Left(edit)).collect(),
        };

        let document_changes = DocumentChanges::Edits(vec![edit]);

        WorkspaceEdit {
            changes: None,
            document_changes: Some(document_changes),
            change_annotations: None,
        }
    } else {
        // Fall back to hash map of `TextEdit`s if the client doesn't support `DocumentChanges`
        let mut changes = HashMap::new();
        changes.insert(uri, edits);

        WorkspaceEdit {
            changes: Some(changes),
            document_changes: None,
            change_annotations: None,
        }
    }
}

impl CodeActions {
    pub(crate) fn new() -> Self {
        Self {
            response: CodeActionResponse::new(),
        }
    }

    pub(crate) fn add_action(&mut self, x: CodeAction) -> Option<()> {
        self.response.push(CodeActionOrCommand::CodeAction(x));
        Some(())
    }

    pub(crate) fn into_response(self) -> CodeActionResponse {
        self.response
    }
}
