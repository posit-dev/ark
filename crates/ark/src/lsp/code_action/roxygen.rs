use tower_lsp::lsp_types;
use url::Url;

use crate::lsp::capabilities::Capabilities;
use crate::lsp::code_action::code_action;
use crate::lsp::code_action::code_action_workspace_text_edit;
use crate::lsp::code_action::CodeActions;
use crate::lsp::documents::Document;
use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::BinaryOperatorType;
use crate::treesitter::NodeTypeExt;

pub(crate) fn roxygen_documentation(
    actions: &mut CodeActions,
    uri: &Url,
    document: &Document,
    range: tree_sitter::Range,
    capabilities: &Capabilities,
) -> Option<()> {
    if !capabilities.code_action_literal_support() {
        // This code action returns literal `CodeAction`s, so must have support for them
        return None;
    }

    // For cursors (the common case), start and end points are the same.
    // For selections, we require that the start point be touching the function name.
    let start = range.start_point;

    let node = document
        .ast
        .root_node()
        .named_descendant_for_point_range(start, start)?;

    // User must be sitting on the function name
    if !node.is_identifier() {
        return None;
    }

    // Parent must be a `<-` or `=` assignment node
    let assignment = node.parent()?;

    if !assignment.is_binary_operator_of_kind(BinaryOperatorType::LeftAssignment) &&
        !assignment.is_binary_operator_of_kind(BinaryOperatorType::EqualsAssignment)
    {
        return None;
    }

    // And the rhs must be a function definition
    let function = assignment.child_by_field_name("rhs")?;

    if !function.is_function_definition() {
        return None;
    }

    // The assignment node must be a direct descendent of the program node (i.e. we don't
    // provide the code action for local functions defined within another function, or
    // within some arbitrary braced expression)
    if !assignment.parent()?.is_program() {
        return None;
    }

    let position = node.start_position();

    // Fairly simple detection of existing `#'` on the previous line (but starting at the
    // same `column` offset), which tells us not to provide this code action
    if let Some(previous_line) = document.contents.get_line(position.row.saturating_sub(1)) {
        if let Some(previous_line) = previous_line.get_byte_slice(position.column..) {
            let mut previous_line = previous_line.bytes();

            if previous_line
                .next()
                .map(|byte| byte == b'#')
                .unwrap_or(false) &&
                previous_line
                    .next()
                    .map(|byte| byte == b'\'')
                    .unwrap_or(false)
            {
                return None;
            }
        }
    }

    // Okay, looks like we are going to provide the code action, collect all parameter
    // names
    let parameters = function.child_by_field_name("parameters")?;

    let mut parameter_names = vec![];
    let mut cursor = parameters.walk();

    for child in parameters.children_by_field_name("parameter", &mut cursor) {
        let parameter_name = child.child_by_field_name("name")?;
        let parameter_name = document
            .contents
            .node_slice(&parameter_name)
            .ok()?
            .to_string();
        parameter_names.push(parameter_name);
    }

    let indent_size = position.column;
    let documentation = documentation_builder(parameter_names, indent_size);

    // We insert the documentation string at the start position of the function name.
    // This handles the indentation of the first documentation line, and makes new line
    // handling trivial (we just add a new line to every documentation line).
    let position = convert_point_to_position(&document.contents, position);
    let range = tower_lsp::lsp_types::Range::new(position, position);
    let edit = lsp_types::TextEdit::new(range, documentation);
    let edit =
        code_action_workspace_text_edit(uri.clone(), document.version, vec![edit], capabilities);

    actions.add_action(code_action(
        "Generate a roxygen template".to_string(),
        lsp_types::CodeActionKind::EMPTY,
        edit,
    ))
}

fn documentation_builder(parameter_names: Vec<String>, indent_size: usize) -> String {
    let mut lines = vec![];

    lines.push("#' Title".to_string());

    if !parameter_names.is_empty() {
        lines.push("#'".to_string());
        lines.append(&mut parameters_builder(parameter_names));
    }

    lines.push("#'".to_string());
    lines.push("#' @returns".to_string());
    lines.push("#'".to_string());
    lines.push("#' @export".to_string());
    lines.push("#' @examples".to_string());

    documentation_from_lines(lines, indent_size)
}

fn parameters_builder(names: Vec<String>) -> Vec<String> {
    names
        .into_iter()
        .map(|name| format!("#' @param {name}"))
        .collect()
}

/// Combine lines into a single documentation string used within a `TextEdit`
///
/// This is done in a clever way:
/// - We don't apply indentation to the first line, as our `TextEdit` will do that for us,
///   by inserting at the specified column position.
/// - We insert a newline after every line, even the last one. This allows us to cleanly
///   insert the documentation at the start position of the function name.
fn documentation_from_lines(lines: Vec<String>, indent_size: usize) -> String {
    let mut documentation = String::new();
    for line in lines {
        documentation.push_str(&line);
        documentation.push('\n');
        documentation.push_str(&" ".repeat(indent_size));
    }
    documentation
}

#[cfg(test)]
mod tests {
    use tower_lsp::lsp_types::CodeActionOrCommand;
    use tower_lsp::lsp_types::DocumentChanges;
    use tower_lsp::lsp_types::OneOf;
    use tower_lsp::lsp_types::Position;
    use tree_sitter::Point;
    use tree_sitter::Range;
    use url::Url;

    use crate::fixtures::point_and_offset_from_cursor;
    use crate::lsp::capabilities::Capabilities;
    use crate::lsp::code_action::roxygen::roxygen_documentation;
    use crate::lsp::code_action::CodeActions;
    use crate::lsp::documents::Document;

    fn point_range(point: Point, byte: usize) -> Range {
        Range {
            start_byte: byte,
            end_byte: byte,
            start_point: point,
            end_point: point,
        }
    }

    fn roxygen_documentation_test(text: &str, position: Position) -> String {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        let capabilities = Capabilities::default()
            .with_code_action_literal_support(true)
            .with_workspace_edit_document_changes(true);

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );

        let mut actions = actions.into_response();
        assert_eq!(actions.len(), 1);

        let CodeActionOrCommand::CodeAction(action) = actions.pop().unwrap() else {
            panic!("Unexpected");
        };
        let workspace_edit = action.edit.unwrap();

        let document_changes = workspace_edit.document_changes.unwrap();
        let DocumentChanges::Edits(mut text_document_edits) = document_changes else {
            panic!("Unexpected");
        };
        assert_eq!(text_document_edits.len(), 1);

        let mut text_document_edit = text_document_edits.pop().unwrap();
        assert_eq!(text_document_edit.text_document.uri, uri);
        assert_eq!(text_document_edit.edits.len(), 1);

        let OneOf::Left(text_edit) = text_document_edit.edits.pop().unwrap() else {
            panic!("Unexpected");
        };
        assert_eq!(text_edit.range.start.line, position.line);
        assert_eq!(text_edit.range.start.character, position.character);
        assert_eq!(text_edit.range.end, text_edit.range.start);

        text_edit.new_text
    }

    #[test]
    fn test_adds_parameters() {
        let new_text = roxygen_documentation_test("fu@n <- function(a, b = 2) {}", Position {
            line: 0,
            character: 0,
        });
        insta::assert_snapshot!(new_text);

        let new_text = roxygen_documentation_test("fu@n <- function(...) {}", Position {
            line: 0,
            character: 0,
        });
        insta::assert_snapshot!(new_text);

        // Mock some new lines and indentation
        // (It's correct for the first line to not be indented in the snapshot,
        // since the `Position` handles the indentation through `character`)
        let new_text = roxygen_documentation_test("\n\n    fu@n <- function(...) {}", Position {
            line: 2,
            character: 4,
        });
        insta::assert_snapshot!(new_text);
    }

    #[test]
    fn test_no_parameters() {
        let new_text = roxygen_documentation_test("fu@n <- function() {}", Position {
            line: 0,
            character: 0,
        });
        insta::assert_snapshot!(new_text);
    }

    #[test]
    fn test_supports_equals_assignment() {
        let new_text = roxygen_documentation_test("fu@n = function(a, b = 2) {}", Position {
            line: 0,
            character: 0,
        });
        insta::assert_snapshot!(new_text);
    }

    #[test]
    fn test_adds_documentation_when_direct_preceding_line_is_not_documentation() {
        let new_text = roxygen_documentation_test("#'\n\nfu@n = function(a, b = 2) {}", Position {
            line: 2,
            character: 0,
        });
        insta::assert_snapshot!(new_text);
    }

    #[test]
    fn test_no_action_when_on_local_function() {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        let capabilities = Capabilities::default()
            .with_code_action_literal_support(true)
            .with_workspace_edit_document_changes(true);

        let text = "
outer <- function(a, b = 2) {
  in@ner <- function(a, b, c) {}
}
        ";

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );
        let actions = actions.into_response();
        assert!(actions.is_empty());
    }

    #[test]
    fn test_no_action_when_documentation_on_previous_line() {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        let capabilities = Capabilities::default()
            .with_code_action_literal_support(true)
            .with_workspace_edit_document_changes(true);

        let text = "
#' Title
f@n <- function(a, b) {}
        ";

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );
        let actions = actions.into_response();
        assert!(actions.is_empty());
    }

    #[test]
    fn test_no_action_when_cursor_is_after_function_name() {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        let capabilities = Capabilities::default()
            .with_code_action_literal_support(true)
            .with_workspace_edit_document_changes(true);

        // This is just how tree-sitter works, it uses a half open range of `[)`.
        let text = "
fn@ <- function(a, b) {}
        ";

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );
        let actions = actions.into_response();
        assert!(actions.is_empty());
    }

    #[test]
    fn test_no_action_without_code_action_literal_support() {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        // NOTE: `with_code_action_literal_support(false)`
        let capabilities = Capabilities::default()
            .with_code_action_literal_support(false)
            .with_workspace_edit_document_changes(true);

        let text = "
f@n <- function(a, b) {}
        ";

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );
        let actions = actions.into_response();
        assert!(actions.is_empty());
    }

    #[test]
    fn test_uses_hash_map_of_text_edits_without_document_changes_support() {
        let mut actions = CodeActions::new();

        let uri = Url::parse("file:///test.R").unwrap();

        // NOTE: `with_workspace_edit_document_changes(false)`
        let capabilities = Capabilities::default()
            .with_code_action_literal_support(true)
            .with_workspace_edit_document_changes(false);

        let text = "
f@n <- function(a, b) {}
        ";

        let (text, point, offset) = point_and_offset_from_cursor(text);
        let document = Document::new(&text, None);

        roxygen_documentation(
            &mut actions,
            &uri,
            &document,
            point_range(point, offset),
            &capabilities,
        );

        let mut actions = actions.into_response();
        assert_eq!(actions.len(), 1);

        let CodeActionOrCommand::CodeAction(action) = actions.pop().unwrap() else {
            panic!("Unexpected");
        };
        let workspace_edit = action.edit.unwrap();

        // This is now `None`
        assert!(workspace_edit.document_changes.is_none());

        // The edits are here instead
        let changes = workspace_edit.changes.unwrap();
        assert_eq!(changes.len(), 1);
        let text_edits = changes.get(&uri).unwrap();
        assert_eq!(text_edits.len(), 1);

        let text_edit = text_edits.get(0).unwrap();
        assert_eq!(text_edit.range.start.line, 1);
        assert_eq!(text_edit.range.start.character, 0);
        assert_eq!(text_edit.range.end, text_edit.range.start);

        insta::assert_snapshot!(text_edit.new_text);
    }
}
