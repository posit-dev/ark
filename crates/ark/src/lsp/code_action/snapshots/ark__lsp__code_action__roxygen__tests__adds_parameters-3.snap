---
source: crates/ark/src/lsp/code_action/roxygen.rs
expression: new_text
---
#' Title
    #'
    #' @param ...
    #'
    #' @returns
    #'
    #' @export
    #' @examples
